### PR TITLE
[MM-39641] Migrate Webhook store to sqlx

### LINKED ISSUE #18879: Migrate from gorp to sqlx in store/sqlstore/webhook_store.go
Mattermost is migrating from using `gorp` as a sql library to using `sqlx`.

This ticket is about migrating all usages of gorp in `store/sqlstore/webhook_store.go` to sqlx.

The way to proceed would be as follows:

All SQL queries are usually executed from `GetMaster()` or `GetReplica()` calls which return the gorp instance for the DB. We have a corresponding method `GetMasterX`/`GetReplicaX` which return the sqlx instance instead.

All named params in SQL query strings should be replaced with `?` and then a list of args should be passed to the method. Here is a quick cheat sheet of the methods to replace with:

```
GetReplica().SelectOne() -> GetReplicaX().Get()
GetReplica().Get() -> GetReplicaX().Get()
GetReplica().Select() -> GetReplicaX().Select()
GetMaster().Insert() -> GetMasterX().NamedExec()
GetMaster().Update() -> GetMasterX().NamedExec()
GetMaster().Exec() -> GetMasterX().Exec()
```

For some practical examples of migrations, please feel free to take a look at:
https://github.com/mattermost/mattermost-server/pull/18302
https://github.com/mattermost/mattermost-server/pull/18301

If you have questions about the ticket or you need any help, feel free to contact `agniva.de.sarker` in https://community.mattermost.com/

----------------------------------

If you're interested please comment here and come [join our "Contributors" community channel](https://community.mattermost.com/core/channels/tickets) on our daily build server, where you can discuss questions with community members and the Mattermost core team. For technical advice or questions, please [join our "Developers" community channel](https://community.mattermost.com/core/channels/developers).

New contributors please see our [Developer's Guide](https://developers.mattermost.com/contribute/getting-started/).

JIRA: https://mattermost.atlassian.net/browse/MM-39641

### PR DESCRIPTION
**Pull request by bretanac93**
For more information, check the ticket [here](https://mattermost.atlassian.net/browse/MM-39641)

Fixes #18879 

```release-note
NONE
```

### CODE REVIEW

File: store/sqlstore/webhook_store.go [hunk 1]
```diff
@@ -86,26 +86,39 @@ func (s SqlWebhookStore) SaveIncoming(webhook *model.IncomingWebhook) (*model.In
...
 	return webhook, nil
-
 }
 
 func (s SqlWebhookStore) UpdateIncoming(hook *model.IncomingWebhook) (*model.IncomingWebhook, error) {
 	hook.UpdateAt = model.GetMillis()
 
-	if _, err := s.GetMaster().Update(hook); err != nil {
+	res, err := s.GetMasterX().NamedExec(`UPDATE IncomingWebhooks SET
+			UpdateAt=:UpdateAt, ChannelId=:ChannelId, TeamId=:TeamId, DisplayName=:DisplayName,
```
Comment (agnivade) [inline store/sqlstore/webhook_store.go:103]: For completeness' sake, let's add `CreateAt` and `DeleteAt` as well.

File: store/sqlstore/webhook_store.go [hunk 2]
```diff
@@ -86,26 +86,39 @@ func (s SqlWebhookStore) SaveIncoming(webhook *model.IncomingWebhook) (*model.In
...
+	if err != nil {
 		return nil, errors.Wrapf(err, "failed to update IncomingWebhook with id=%s", hook.Id)
 	}
+	count, err := res.RowsAffected()
+	if err != nil {
+		return nil, errors.Wrap(err, "error while getting rows_affected")
+	}
+	if count > 1 {
+		return nil, store.NewErrInvalidInput("IncomingWebhook", "Id", hook.Id)
+	}
```
Comment (agnivade) [inline store/sqlstore/webhook_store.go:115]: There is no need for these since they weren't already there.

File: store/sqlstore/webhook_store.go [hunk 3]
```diff
@@ -163,7 +176,7 @@ func (s SqlWebhookStore) GetIncomingListByUser(userId string, offset, limit int)
 		return nil, errors.Wrap(err, "incoming_webhook_tosql")
```
Comment (agnivade) [inline store/sqlstore/webhook_store.go:176]: In L163, change `webhooks := []*model.IncomingWebhook{}`

File: store/sqlstore/webhook_store.go [hunk 4]
```diff
@@ -191,7 +204,7 @@ func (s SqlWebhookStore) GetIncomingByTeamByUser(teamId string, userId string, o
...
```
Comment (agnivade) [inline store/sqlstore/webhook_store.go:204]: Same comment as above.

File: store/sqlstore/webhook_store.go [hunk 5]
```diff
@@ -205,7 +218,7 @@ func (s SqlWebhookStore) GetIncomingByTeam(teamId string, offset, limit int) ([]
 func (s SqlWebhookStore) GetIncomingByChannel(channelId string) ([]*model.IncomingWebhook, error) {
 	var webhooks []*model.IncomingWebhook
```
Comment (agnivade) [inline store/sqlstore/webhook_store.go:219]: `webhooks := []*model.IncomingWebhook{}`

File: store/sqlstore/webhook_store.go [hunk 6]
```diff
@@ -263,7 +281,7 @@ func (s SqlWebhookStore) GetOutgoingListByUser(userId string, offset, limit int)
 		return nil, errors.Wrap(err, "outgoing_webhook_tosql")
 	}
```
Comment (agnivade) [inline store/sqlstore/webhook_store.go:283]: Same comment as earlier.

File: store/sqlstore/webhook_store.go [hunk 7]
```diff
@@ -298,7 +316,7 @@ func (s SqlWebhookStore) GetOutgoingByChannelByUser(channelId string, userId str
...
```
Comment (agnivade) [inline store/sqlstore/webhook_store.go:318]: Same comment as earlier.

File: store/sqlstore/webhook_store.go [hunk 8]
```diff
@@ -332,7 +350,7 @@ func (s SqlWebhookStore) GetOutgoingByTeamByUser(teamId string, userId string, o
...
```
Comment (agnivade) [inline store/sqlstore/webhook_store.go:352]: Same comment as earlier.

File: store/sqlstore/webhook_store.go [hunk 9]
```diff
@@ -375,51 +393,66 @@ func (s SqlWebhookStore) PermanentDeleteOutgoingByChannel(channelId string) erro
...
 		return nil, errors.Wrapf(err, "failed to update OutgoingWebhook with id=%s", hook.Id)
 	}
-
+	count, err := res.RowsAffected()
+	if err != nil {
+		return nil, errors.Wrap(err, "error while getting rows_affected")
+	}
+	if count > 1 {
+		return nil, store.NewErrInvalidInput("OutgoingWebhook", "Id", hook.Id)
+	}
```
Comment (agnivade) [inline store/sqlstore/webhook_store.go:410]: We can remove these as well.

File: store/sqlstore/webhook_store.go [hunk 10]
```diff
@@ -375,51 +393,66 @@ func (s SqlWebhookStore) PermanentDeleteOutgoingByChannel(channelId string) erro
 func (s SqlWebhookStore) UpdateOutgoing(hook *model.OutgoingWebhook) (*model.OutgoingWebhook, error) {
 	hook.UpdateAt = model.GetMillis()
 
-	if _, err := s.GetMaster().Update(hook); err != nil {
+	res, err := s.GetMasterX().NamedExec(`UPDATE OutgoingWebhooks SET
```
Comment (agnivade) [inline store/sqlstore/webhook_store.go:396]: Let's add `CreateAt` and `DeleteAt` as well.

File: store/sqlstore/webhook_store.go [hunk 11]
```diff
@@ -375,51 +393,66 @@ func (s SqlWebhookStore) PermanentDeleteOutgoingByChannel(channelId string) erro
...
 func (s SqlWebhookStore) AnalyticsIncomingCount(teamId string) (int64, error) {
-	query :=
-		`SELECT
-			COUNT(*)
-		FROM
-			IncomingWebhooks
-		WHERE
-			DeleteAt = 0`
+	queryBuilder :=
+		s.getQueryBuilder().
```
Comment (agnivade) [inline store/sqlstore/webhook_store.go:416]: Thank you for converting them to squirrel! :heart:

File: store/sqlstore/webhook_store.go [hunk 12]
```diff
@@ -375,51 +393,66 @@ func (s SqlWebhookStore) PermanentDeleteOutgoingByChannel(channelId string) erro
...
 func (s SqlWebhookStore) AnalyticsOutgoingCount(teamId string) (int64, error) {
-	query :=
-		`SELECT
-			COUNT(*)
-		FROM
-			OutgoingWebhooks
-		WHERE
-			DeleteAt = 0`
+	queryBuilder :=
+		s.getQueryBuilder().
```
Comment (agnivade) [inline store/sqlstore/webhook_store.go:439]: :heart:

**Review by agnivade — COMMENTED** [review]
Great stuff @bretanac93 ! Some small suggestions to take care of.

File: store/sqlstore/webhook_store.go (hunk 1 above)
Comment (bretanac93) [reply store/sqlstore/webhook_store.go]: Fixed in 9ebd5c7df

File: store/sqlstore/webhook_store.go (hunk 2 above)
Comment (bretanac93) [reply store/sqlstore/webhook_store.go]: Fixed in 9ebd5c7df

**Review by agnivade — APPROVED** [review]
Great stuff!

**Comment by emilyacook** [comment]
Thanks for participating in Hacktoberfest! You can claim your sticker set here: https://get.printfection.com/hacktober21/4144583267 @bretanac93

**Review by isacikgoz — APPROVED** [review]
LGTM! Thanks for the contribution @bretanac93 🚀

**Pull request merged**